### PR TITLE
feat: Move probe builders to ProbeGenerator class

### PR DESCRIPTION
**Pull request by samuel-hawker**
Probe builders for now belong in abstractmodel
This will allow for non static referencing in future
changes.
ModelUtils whilst useful is an antipattern that encourages
large code groupings.

Signed-off-by: Samuel Hawker <samuel.hawker@ibm.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by scholzj** [comment]
TBH, I think we intentionally created the ModelUtils class exactly for the purpose to make the AbstractModel more lightweight and not to have there so many different methods which do not really need to be there and can be somewhere else as static methods.

**Comment by samuel-hawker** [comment]
I can see your point here, but putting it in a class as generic as ModelUtils is also just the same as this.
Would you be against me making a new class called:
`ProbeGenerator`

**Comment by samuel-hawker** [comment]
Updated PR to reflect new pattern that I think is far more desirable and encourages code re-use.

**Comment by scholzj** [comment]
Sorry to be a pita ... but I still don't understand why is this better that just a class with static methods? The code seems much harder to read to me and I do not see any added value. The mixture of different patterns and styles is also very confusing to me ... There is a builder, there is a constructor, there is a static initialiser, there are completely static methods ... then sometimes you use one and sometimes the other without any obvious key.

**Comment by samuel-hawker** [comment]
I think you raise some very valid points, that I hope I address in the following:

As a general point, I am attempting to consolidate the ModelUtils methods down into separate packages where they logically belong. Right now ModelUtils is sort of a place where anything that doesn't clearly belong a pre-existing class then it is put there, I think this generally leads to bad coding habits, such as messy code, unknown duplication etc and to this end I am writing new classes for these functions and refactoring a bit with it.

- why is this better that just a class with static methods?
Given the current approach of having one single type of liveness/readiness it isn't much better, but going forward if we were to support say, both https + https liveness probes based on configuration, this makes working the code together easier.
It is also easier, in a separate class to keep track of whether it has any tests (currently these methods don't have any direct tests)

- The code seems much harder to read to me and I do not see any added value.
I have isolated the common logic (the options setting) and put it in the `of` call which is a shorthand for calling the builder, with any validation, we may optionally want to add later.

- The mixture of different patterns and styles is also very confusing to me
I am interested which patterns you find confusing, do you not like the 
`Generator.of(...).createThing()` pattern, this is very similar to the patterns we are using currently with the models, it is a form of factory?

- There is a builder, there is a constructor, there is a static initialiser, there are completely static methods ... then sometimes you use one and sometimes the other without any obvious key.
I think the TlsSidecar stuff is the most confusing, I decided to keep those as static as they seem to have a very rigid flow and get re-used in multiple places it didn't seem worth it to switch those over to nonstatic

As a general point about static initialisers, I have been watching quite a few talks on clean-code about them, which speak of how they're useful for checking for invalid provided config, the aim was to reduce the amount of nulls that we are handling.

**Comment by scholzj** [comment]
> Given the current approach of having one single type of liveness/readiness it isn't much better, but going forward if we were to support say, both https + https liveness probes based on configuration, this makes working the code together easier.
It is also easier, in a separate class to keep track of whether it has any tests (currently these methods don't have any direct tests)

I do not think I have any problem with the separate class. I can understand why having it in a separate class makes sense. It is more the content of the class I'm not sure about.

> I am interested which patterns you find confusing, do you not like the
Generator.of(...).createThing() pattern, this is very similar to the patterns we are using currently with the models, it is a form of factory?

I do not particularly like the `Generator.of(...).createThing()`. I wonder what is for example the impact on memory from creating an object instance instead of using static method (IMHO very important given our increasing memory demands). But on its own it is not confusing. Confusing is why you use it somewhere but use static methods somewhere else.

> I think the TlsSidecar stuff is the most confusing, I decided to keep those as static as they seem to have a very rigid flow and get re-used in multiple places it didn't seem worth it to switch those over to nonstatic

Right, this is the most confusing part. You are suggesting that you are introducing a better way. But then for half of it you don't use it.

**Comment by ppatierno** [comment]
I totally agree with having a separate class. Having one "bucket" to put a ton of methods calling them just "utils" is something I don't like. Grouping methods by purpose is really useful.
Regarding the way, I don't understand the need for a factory creating every time a new instance of `ProbeGenerator` instead of having static methods.

**Comment by samuel-hawker** [comment]
Thanks for the feedback both of you.

I think it is clear to me now my builder structure was a misguided design, I will switch back to using static callers.

**Comment by samuel-hawker** [comment]
Thanks for the feedback, I am actually much happier with this new implementation, it has best of both worlds, plus added tests to ensure that it all works as intended.

**Comment by samuel-hawker** [comment]
To address Jakub's comments about memory, since the fields are immutable static final, the overhead of initializing them in memory, i believe is very small.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ProbeGenerator.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    private ProbeGenerator(Probe probeConfig) {
+        this.probeConfig = probeConfig;
+    }
+
+    public static ProbeGenerator of(Probe probeConfig) {
+        if (probeConfig == null) {
+            throw new IllegalArgumentException();
+        }
+        return new ProbeGenerator(probeConfig);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ProbeGenerator.java:26]: This is really what my concern with regards to the memory was. If you are sure it has no impact I'm fine. I admit that  I have no clue how things like memory work in Java or any other language :-o.

**Review by scholzj — APPROVED** [review]
I guess I can live with this.

**Comment by samuel-hawker** [comment]
Any thoughts on this @ppatierno

**Comment by ppatierno** [comment]
@samuel-hawker I have still concerns about using the static methods but every time creating a new `ProbeGenerator` instance that has a final field which is not static right? So it's allocated every time. Why do we need this?

**Comment by samuel-hawker** [comment]
I think you're right that there is not enough value in having this have a constructor, I will remove the fields to drop the allocations.

**Review by ppatierno — APPROVED** [review]

**Pull request merged**